zhaquirks/sonoff/trvzb.py

```diff
@@ -0,0 +1,116 @@
+"""Device handler for Sonoff TRVZB"""
+import logging
+import typing
+
+from zigpy.profiles import zha
+from zigpy.quirks import CustomCluster, CustomDevice
+import zigpy.types as t
+from zigpy.zcl.clusters.general import (
+    Basic,
+    Identify,
+    OnOff,
+    Ota,
+    PollControl,
+    PowerConfiguration,
+    Time,
+)
+from zigpy.zcl.clusters.hvac import Thermostat
+from zigpy.zcl.foundation import BaseAttributeDefs, ZCLAttributeDef
+
+from zhaquirks.const import (
+    DEVICE_TYPE,
+    ENDPOINTS,
+    INPUT_CLUSTERS,
+    MODELS_INFO,
+    OUTPUT_CLUSTERS,
+    PROFILE_ID,
+)
+
+LOGGER = logging.getLogger(f"zhaquirks.{__name__}")
+
+SONOFF_CLUSTER_ID: typing.Final = 0xFC11
+SONOFF_CLUSTER_ID_FC57: typing.Final = 0xFC57
+SONOFF_ATTR_CHILD_LOCK: typing.Final = 0x0000
+SONOFF_ATTR_WINDOW_OPEN: typing.Final = 0x6000
+
+
+class SonoffManufCluster(CustomCluster):
+    """Sonoff manufacturer specific cluster."""
+
+    name: str = "Sonoff Manufacturer Specific"
+    cluster_id: t.uint16_t = SONOFF_CLUSTER_ID
+    ep_attribute: str = "sonoff_manufacturer"
+
+    class AttributeDefs(BaseAttributeDefs):
+        child_lock: typing.Final = ZCLAttributeDef(
+            id=SONOFF_ATTR_CHILD_LOCK,
+            type=t.Bool,
+            mandatory=True,
+            is_manufacturer_specific=False,
+        )
+        open_window: typing.Final = ZCLAttributeDef(
+            id=SONOFF_ATTR_WINDOW_OPEN,
+            type=t.Bool,
+            mandatory=True,
+            is_manufacturer_specific=False,
+        )
+
+    @property
+    def _is_manuf_specific(self) -> bool:
+        """Pretend this is not manufacturer specific as the device will always
+        return Status.UNSUPPORTED_ATTRIBUTE if the manufacturer id is set
+        """
+        return False
```

Comment:
So both changing `is_manufacturer_specific` to `False` for the attribute and overriding `_is_manuf_specific` (which probably shouldn't be done) still causes zigpy/ZHA to send the manufacturer attribute code thus causing the Sonoff to return "unsupported attribute"?

Reply:
Yes, exactly. Reading he child lock via HA's "Manage Zigbee decive" dialog does still set `is_manufacturer_specific=True` as well as `is_manufacturer_specific=True`:

```
2023-12-13 09:31:13.840 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Sending request header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=True, direction=<Direction.Server_to_Client: 0>, disable_default_response=0, reserved=0, *is_cluster=False, *is_general=True), manufacturer=4742, tsn=15, command_id=<GeneralCommand.Read_Attributes: 0>, *direction=<Direction.Server_to_Client: 0>)
2023-12-13 09:31:13.840 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Sending request: Read_Attributes(attribute_ids=[0])
2023-12-13 09:31:13.841 DEBUG (MainThread) [zigpy.device] [0xd636] Extending timeout for 0x0f request
2023-12-13 09:31:15.550 DEBUG (MainThread) [zigpy.application] Received a packet: ZigbeePacket(timestamp=datetime.datetime(2023, 12, 13, 8, 31, 15, 550115, tzinfo=datetime.timezone.utc), src=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0xD636), src_ep=1, dst=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0x0000), dst_ep=1, source_route=None, extended_timeout=False, tsn=0, profile_id=260, cluster_id=64529, data=Serialized[b'\x1c\x86\x12\x0f\x01\x00\x00\x86'], tx_options=<TransmitOptions.NONE: 0>, radius=29, non_member_radius=0, lqi=196, rssi=None)
2023-12-13 09:31:15.551 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Received ZCL frame: b'\x1c\x86\x12\x0f\x01\x00\x00\x86'
2023-12-13 09:31:15.551 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Decoded ZCL frame header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=True, direction=<Direction.Client_to_Server: 1>, disable_default_response=1, reserved=0, *is_cluster=False, *is_general=True), manufacturer=4742, tsn=15, command_id=1, *direction=<Direction.Client_to_Server: 1>)
2023-12-13 09:31:15.552 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Decoded ZCL frame: SonoffManufCluster:Read_Attributes_rsp(status_records=[ReadAttributeRecord(attrid=0x0000, status=<Status.UNSUPPORTED_ATTRIBUTE: 134>)])
2023-12-13 09:31:15.553 DEBUG (MainThread) [homeassistant.components.zha.websocket_api] Read attribute for: cluster_id: [64529] cluster_type: [in] endpoint_id: [1] attribute: [0] manufacturer: [4742] response: [None] failure: [{0: <Status.UNSUPPORTED_ATTRIBUTE: 134>}],
```


vs:

```
2023-12-13 09:34:31.991 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Sending request header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=False, direction=<Direction.Server_to_Client: 0>, disable_default_response=0, reserved=0, *is_cluster=False, *is_general=True), tsn=16, command_id=<GeneralCommand.Read_Attributes: 0>, *direction=<Direction.Server_to_Client: 0>)
2023-12-13 09:34:31.991 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Sending request: Read_Attributes(attribute_ids=[0])
2023-12-13 09:34:31.991 DEBUG (MainThread) [zigpy.device] [0xd636] Extending timeout for 0x10 request
2023-12-13 09:34:32.267 DEBUG (MainThread) [zigpy.application] Received a packet: ZigbeePacket(timestamp=datetime.datetime(2023, 12, 13, 8, 34, 32, 267321, tzinfo=datetime.timezone.utc), src=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0xD636), src_ep=1, dst=AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0x0000), dst_ep=1, source_route=None, extended_timeout=False, tsn=0, profile_id=260, cluster_id=64529, data=Serialized[b'\x18\x10\x01\x00\x00\x00\x10\x00'], tx_options=<TransmitOptions.NONE: 0>, radius=29, non_member_radius=0, lqi=196, rssi=None)
2023-12-13 09:34:32.267 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Received ZCL frame: b'\x18\x10\x01\x00\x00\x00\x10\x00'
2023-12-13 09:34:32.268 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Decoded ZCL frame header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=0, direction=<Direction.Client_to_Server: 1>, disable_default_response=1, reserved=0, *is_cluster=False, *is_general=True), tsn=16, command_id=1, *direction=<Direction.Client_to_Server: 1>)
2023-12-13 09:34:32.268 DEBUG (MainThread) [zigpy.zcl] [0xD636:1:0xfc11] Decoded ZCL frame: SonoffManufCluster:Read_Attributes_rsp(status_records=[ReadAttributeRecord(attrid=0x0000, status=<Status.SUCCESS: 0>, value=TypeValue(type=Bool, value=<Bool.false: 0>))])
2023-12-13 09:34:32.269 DEBUG (MainThread) [homeassistant.components.zha.websocket_api] Read attribute for: cluster_id: [64529] cluster_type: [in] endpoint_id: [1] attribute: [0] manufacturer: [None] response: [Bool.false] failure: [{}],
```

Reply:
> overriding `_is_manuf_specific` (which probably shouldn't be done) ...

AIUI `_has_manuf_arttr` will always return true as long as `_is_manuf_specific` is true (e.g. without checking the actual attributes). That's why I've tried it that way.